src/messages/Lifecycle.ts

```diff
@@ -0,0 +1,64 @@
+import { guard, mixed, object, string } from 'decoders';
+import { LabeledMsg } from './LabeledMsg';
+import { createMessageValidator } from './validationUtils';
+
+/**
+ * Client started indication.  The client will
+ * now listen for messages and begin sending
+ * messages from the client application.
+ */
+export interface LabeledStarted extends LabeledMsg {
+  msgType: 'client_started';
+}
+
+// We don't care what is in msg for Started messages.
+const startedDecoder = guard(mixed);
+
+const validateStarted = createMessageValidator<LabeledStarted>(
+  'client_started',
+  startedDecoder
+);
+
+/**
+ * Environmental data provided to all clients
+ * in order to match behavior of the host application.
+ */
+export interface EnvData {
+  locale: string;
+}
```

Comment:
I would add `hostRootUrl` here. There's a lot of useful information that can be derived from the client knowing who's hosting it. Also, it will enable the client to generate links that work correctly when opened in a new tab or shared via email/chat in the cases where URL routing is used.

Reply:
I think this is an API that would benefit from a `custom: any` field that could pass through arbitrary application-specific data. That's actually where I'd handle `platformId`, for a couple of reasons. One is that I'd like to keep the library geared toward general-purpose use and not encode too much that's specific to Genesys into it. The other is that we may end up using this in some unexpected cases for things like internal tooling where `platformId` doesn't make sense, or would end up growing in a way that might not be useful. 

What's the logical difference between locale and language in this case?  In PureCloud we don't separate the two, so I suspect this is a feature difference between platforms that I'm just not expecting.

Reply:
I'll update to use a custom.  Of course we won't be able to validate those data, so I would probably discourage using that as much as possible since its an implicit API.

Language is used to determine the translations the UI strings use.  Locale controls how certain things are formatted (numbers [3,2 vs 3.2], currency, and datetimes).  It also is supposed to control how things are sorted, but that doesn't work great in JS yet.
Separating these two allows you to write fewer translation packs while still working in many regions (de-DE vs de-AT for example).

Also in theory you might be forced to use an english translation, but be located in another country (e.g. Netherlands) and wish to display the values in the location.

Reply:
As @MattCheely mentioned, we usually just let the user select their preferred language and that is represented in a single value, `userPrefLangTag`.  This is in the format of a BCP 47 lang tag.  This allows us to pick the best supported language bundle we have for the user (dropping dialects) and also allows fallback based on the user agent's language property which is also in BCP 47.  Also, bcp 47 format is supported by the JS intl api.

Perhaps, we should be more specfic either in property name or type to indicate which type of value is expected here.  We've got a lot of ugly legacy code transforming all manor of language identifiers from our earlier days :(

+1 on separating language from other concerns.  I'm just not sure locale goes far enough or is the most obvious label.  Perhaps we could be more specific?  Something like dateFormatLangTag, numberFormatLangTag, phoneNumberLangTag?  All of these could fall back to the default langTag if not specified.  At some point, we'll have to cut it off and move into the custom section.

Reply:
In PureConnect, we use separate codes.  We default their display language to the browser language if we can as well, but that may not always be the choice they want to use for the displayed language (there may be a native Spanish speaker in an english call center).  And we always use their browser setting for their locale (used with JS Intl and a couple of our third-party dists).

Has anyone talked with Localization what their plans are?  That would certainly help us decide this I think.

Also I really don't think we need to have more specific tags.  It's extremely rare that you would use one locale for numbers and another for dates. As for naming... .NET calls these CommonCulture and CommonUICulture (https://blogs.msdn.microsoft.com/heaths/2005/02/17/the-differences-between-locales-and-languages/).  In this draft WEC document they used 'Langauge Tag' and 'Locale Identifier': https://www.w3.org/TR/ltli/.

Reply:
Here are the options I see for language/locale stuff:

1. Keep it as is with `language` and `locale`. That lets us handle scenarios where language is different from locale, which is a nice feature if the platform supports it. Where they're the same, we can just use the same value. It doesn't provide for more complex customization options, although at that point, I think we'd want apps to make an API call or use the `custom` field anyway. I do worry a little bit that in platforms where they're not different, it's not clear which one is the source of truth (they shouldn't be different, but mistakes happen).
2. Just have `locale` since it's the _minimum_ necessary to render, and defer any additional options to `custom` on a per-platform basis.
3. Decide the problem isn't well understood right now, take them both out, and rely on `custom` until some consistent pattern emerges.

Personally, I prefer options 2 and 3 but the API as it is now isn't a deal-breaker, IMO.